apps/aecore/src/aec_db_gc.erl

```diff
@@ -0,0 +1,227 @@
+             #data{enabled = true, synced = true, height = LastHeight, hashes = Hashes} = Data)
+  when is_reference(Hashes), Height > LastHeight ->
+    {ok, _} = range_collect_reachable_hashes(Height, Data),
+    {keep_state, Data#data{height = Height}};
+
+handle_event({call, From}, maybe_garbage_collect, ready,
+             #data{enabled = true, synced = true, hashes = Hashes} = Data)
+  when Hashes /= undefined, not is_pid(Hashes) ->
+    Header = aec_chain:top_header(),
+    case aec_headers:type(Header) of
+        key ->
+            Height  = aec_headers:height(Header),
+            {ok, _} = range_collect_reachable_hashes(Height, Data),
+            {ok, N} = swap_nodes(Hashes),
+            ets:delete(Hashes),
+            {next_state, idle, Data#data{height = undefined, hashes = undefined},
+             {reply, From, {ok, N}}};
+        micro ->
+            {keep_state, Data}
+    end;
+handle_event({call, From}, maybe_garbage_collect, _, Data) ->
+    {keep_state, Data, {reply, From, nop}};
+
+handle_event(_, _, _, Data) ->
+    {keep_state, Data}.
+
+
+terminate(_Reason, _State, _Data) -> void.
+
+code_change(_, State, Data, _) -> {ok, State, Data}.
+
+callback_mode() -> handle_event_function.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+%% From - To (inclusive)
+collect_reachable_hashes(FromHeight, ToHeight) when FromHeight < ToHeight ->
+    {ok, Hashes} = collect_reachable_hashes_fullscan(FromHeight),
+    {ok, Hashes} = range_collect_reachable_hashes(FromHeight, ToHeight, Hashes),
+    {ok, Hashes}.
+
+collect_reachable_hashes_fullscan(Height) ->
+    Tab = ets:new(gc_reachable_hashes, [public]), %% TODO: remove public
+    MPT = get_mpt(Height),
+    ?LOG("scanning accounts full tree with root hash ~w at height = ~p~n",
+         [aeu_mp_trees:root_hash(MPT), Height]),
+    {ok, aeu_mp_trees:visit_reachable_hashes(MPT, Tab, fun store_hash/3)}.
+
+%% assumes Height - 1, Height - 2, ... down to Height - History
+%% are in Hashes from previous runs
+collect_reachable_hashes_delta(Height, Hashes) ->
+    MPT = get_mpt(Height),
+    ?LOG("scanning accounts changes with root hash ~w at height = ~p~n",
+         [aeu_mp_trees:root_hash(MPT), Height]),
+    {ok, aeu_mp_trees:visit_reachable_hashes(MPT, Hashes, fun store_unseen_hash/3)}.
+
+range_collect_reachable_hashes(ToHeight, #data{height = LastHeight, hashes = Hashes}) ->
+    range_collect_reachable_hashes(LastHeight, ToHeight, Hashes).
+range_collect_reachable_hashes(LastHeight, ToHeight, Hashes) ->
+    [collect_reachable_hashes_delta(H, Hashes) || H <- lists:seq(LastHeight + 1, ToHeight)],
+    {ok, Hashes}.
+
+%% the actual GC: clear whole table + insert reachable nodes
+swap_nodes(Hashes) ->
+    NodesCount = ets:info(Hashes, size),
+    %% clearing tab can't run in transaction
```

Comment:
It can, if you do something like this:
```
mnesia:transaction(fun() ->
                        {_Mod,Tid,Ts} = get(mnesia_activity_state),
                        mnesia:clear_table(Tid,Ts,Tab,'_')
                   end).
```

Reply:
That looks very helpful, going to try that!

Reply:
Those lovely Mnesia internal APIs :-)